stdlib/toolchain/Compatibility51/ProtocolConformance.cpp

```diff
@@ -0,0 +1,677 @@
+    return nullptr;
+      
+  case TypeReferenceKind::DirectObjCClassName:
+    if (auto cls = reinterpret_cast<const ClassMetadata *>(
+                             objc_lookUpClass(conf->getDirectObjCClassName())))
+      return getObjCClassMetadata(cls);
+    return nullptr;
+
+  case TypeReferenceKind::DirectTypeDescriptor:
+  case TypeReferenceKind::IndirectTypeDescriptor: {
+    if (auto anyType = conf->getTypeDescriptor()) {
+      if (auto type = dyn_cast<TypeContextDescriptor>(anyType)) {
+        if (!type->isGeneric()) {
+          if (auto accessFn = type->getAccessFunction())
+            return accessFn(MetadataState::Abstract).Value;
+        }
+      } else if (auto protocol = dyn_cast<ProtocolDescriptor>(anyType)) {
+        auto protocolRef = ProtocolDescriptorRef::forSwift(protocol);
+        auto constraint =
+          protocol->getProtocolContextDescriptorFlags().getClassConstraint();
+        return getExistentialTypeMetadata(constraint,
+                                          /*superclass bound*/ nullptr,
+                                          /*num protocols*/ 1,
+                                          &protocolRef);
+      }
+    }
+
+    return nullptr;
+  }
+  }
+
+  swift_runtime_unreachable("Unhandled TypeReferenceKind in switch.");
+}
+
+  class ConformanceCandidate {
+    const void *candidate;
+    bool candidateIsMetadata;
+
+  public:
+    ConformanceCandidate() : candidate(0), candidateIsMetadata(false) { }
+
+    ConformanceCandidate(const ProtocolConformanceDescriptor &conformance)
+      : ConformanceCandidate()
+    {
+      if (auto description = conformance.getTypeDescriptor()) {
+        candidate = description;
+        candidateIsMetadata = false;
+        return;
+      }
+
+      if (auto metadata = override_getCanonicalTypeMetadata(&conformance)) {
+        candidate = metadata;
+        candidateIsMetadata = true;
+        return;
+      }
+    }
+
+    /// Retrieve the conforming type as metadata, or NULL if the candidate's
+    /// conforming type is described in another way (e.g., a nominal type
+    /// descriptor).
+    const Metadata *getConformingTypeAsMetadata() const {
+      return candidateIsMetadata ? static_cast<const Metadata *>(candidate)
+                                 : nullptr;
+    }
+
+    const ContextDescriptor *
+    getContextDescriptor(const Metadata *conformingType) const {
+      const auto *description = conformingType->getTypeContextDescriptor();
+      if (description)
+        return description;
+
+      // Handle single-protocol existential types for self-conformance.
+      auto *existentialType = dyn_cast<ExistentialTypeMetadata>(conformingType);
+      if (existentialType == nullptr ||
+          existentialType->getProtocols().size() != 1 ||
+          existentialType->getSuperclassConstraint() != nullptr)
+        return nullptr;
+
+      auto proto = existentialType->getProtocols()[0];
+
+#if SWIFT_OBJC_INTEROP
```

Comment:
Maybe drop this check, since the compatibility library is only built for platforms where this is true. (Unless you'd rather keep as close to the original code as possible, which would be a reasonable goal.)